Fix: Improve Custom MCP node performance by caching toolkits

toi500 opened the pull request:
As discussed, this PR fixes a critical performance issue in the `Custom MCP` node caused by redundant process spawning. The STDIO transport is re-spawning a child process on every UI refresh, leading to significant overhead.

**Solution:**

This is a single file solution implemented entirely within `customMCP.ts`.

1.  A `static toolkitCache` is used to store and reuse `MCPToolkit` instances, preventing the expensive process creation on every refresh.
2.  The cache key is generated by parsing the `mcpServerConfig` JSON. This makes the cache immune to formatting differences (spaces, newlines) and prevents spawning duplicate processes for the same logical configuration.

![chrome_6CVrqvDHHL](https://github.com/user-attachments/assets/eb082f4a-d1a5-4700-b615-48829bdf76d7)


**Trade-off:**

Since we are not modifying `core.ts` and the node's `destroy()` method has proven unreliable for cleanup (I have tried to implement it without any success), this solution will leak one process per unique server configuration. This is a necessary trade-off to fix the critical performance issue immediately while keeping the changes isolated to this single file.

toi500 commented:
@HenryHengZJ, this is working very well for me now, very smooth.

I really hope you can find a solution to flush the cache when the custom node is deleted. I tried to implement this myself but without any success.

I will definitely be keeping this change for my own use, as it is much better to have a minor memory leak (especially since the current MCP implementation has some leaks anyway) than to have the constant process churn overload my server and force me to spend 20 minutes filling in actions on a single MCP node.

toi500 commented:
I just tested the memory leak in production for this PR, and it's basically the same as the current MCP implementation (without this PR) -> https://github.com/FlowiseAI/Flowise/issues/4369#issuecomment-2851624658

Using this PR:

![image](https://github.com/user-attachments/assets/879a6ec4-f3ce-417f-a486-298424af77c7)

HenryHengZJ commented on packages/components/nodes/tools/MCP/CustomMCP/CustomMCP.ts:
```diff
@@ -140,6 +143,25 @@ class Custom_MCP implements INode {
+            // If parsing fails (e.g., invalid JSON), use the raw string as a fallback.
+            canonicalConfig = mcpServerConfig
+        }
+        const cacheKey = JSON.stringify({ workspaceId, canonicalConfig, sandbox })
```
@toi500 im quite concerned about stringify the whole canonicalConfig and sandboxVars because users can enter anything that could cause problem.

all we need is a unique key right?
how bout `workspaceId_chatflowId_nodeId` ?

toi500 replied:
This is the good news is that don't need to clear any cache any longer. Currently, when you set an MCP server config, that service will keep running in the background until you reboot the Flowise container, even if you delete the MCP tools or the flow. The cache is cleared the same way, when the Flowise instance is rebooted.

We could clear the cache in several ways, but if the associated service keeps running in the background, it would be pointless.

ahmedrowaihi commented:
does this takes care of tool availability changes ? when custom mcps are used via `npx` the mcp tools might change ( which is desired if user wants new version of that mcp server ), but I wonder this cache will not affect that ?

toi500 commented:
<img width="840" height="1247" alt="image" src="https://github.com/user-attachments/assets/a9d9b0bb-fe43-4d4a-81c9-e9f7f0fbdebb" />

toi500 commented:
@HenryHengZJ, this PR is ready to review. I have tested it properly, and it works in every edge scenario I thought of. I added it as an option so it may solve any extreme scenario.

Since you pointed out that this cache approach is not compatible with `Flowise Cloud `and it makes sense more for `self-hosted` users only, I have also added a flag for queue mode, so on MODE=queue the cache is to be auto-disabled and the UI option hidden.

If you test this on queue mode, the node should work as it does right now (no cache involved, no "Disable Cache" option visible).

HenryHengZJ commented:
@toi500 I've been thinking about this and trying to see if I can find an optimal way for both cloud and self hosted users.

I found a better solution for self hosted:
1.) We already have a centralize cache called `CachePool` which we can reuse
2.) I've hashed the workspaceId and canonicalConfig as the key of the cache, so we can reduce the byte size

For flowise cloud users in QUEUE mode, unfortunately we can't store the toolkit classes into Redis, so its not achievable.

Can you try this branch: https://github.com/FlowiseAI/Flowise/tree/bugfix/Custom-MCP-Cache

I've tried multiple times and it works

toi500 commented:
@HenryHengZJ, that makes sense. I just tried it, and it works for me too.

HenryHengZJ commented:
PR opened here: https://github.com/FlowiseAI/Flowise/pull/4897
